### PR TITLE
Tests and minor improvements 🎉 

### PR DESCRIPTION
**Pull request by pedrobslisboa**
Laura, I took the liberty to start testing the library, if there is no problem with it, follow the change bellow :)

#### Changes proposed in this Pull Request:

- Creation of tests (About test reducer https://testing-library.com/docs/example-react-hooks-usereducer/)
- Minor change on `initialItems` type
- Turn off eslint error for dev dependencies

#### Testing instructions:

As it is a test construction you can test it running yarn test.
You can also run `yarn test --coverage` to check it coverage

#### Proposed changelog entry for your changes:

- Use Selected Items Hook: set initialItems as optional since it has a default value `[]`
- Tests: add test to library

### CODE REVIEW

**Comment by pedrobslisboa** [comment]
Coverage:
Screenshot 2021-02-05 at 10 20 56

**Comment by LauraBeatris** [comment]
Hey @pedrobslisboa 

Thanks for your contribution 😄
I'll review the PR as soon as possible.

File: package.json [hunk 1]
```diff
@@ -61,6 +61,8 @@
     "lint-staged": "^10.5.1",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
+    "react-hooks-testing-library": "^0.6.0",
```
Comment (LauraBeatris) [inline package.json]: `react-hooks-testing-library` is deprecated. Refer to https://www.npmjs.com/package/react-hooks-testing-library?activeTab=readme#-deprecated---this-package-has-moved-

`@testing-library/react-hooks` should be installed instead

File: .eslintrc.js [hunk 2]
```diff
@@ -9,15 +9,13 @@ module.exports = {
     "@typescript-eslint/explicit-module-boundary-types": "off",
     "@typescript-eslint/explicit-function-return-type": "off",
     "@typescript-eslint/no-explicit-any": "off",
+    "import/no-extraneous-dependencies": "off",
```
Comment (LauraBeatris) [inline .eslintrc.js]: This rule is important to avoid importing packages that aren't declared in the `package.json` dependencies.

Refer to https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md#importno-extraneous-dependencies-forbid-the-use-of-extraneous-packages

And `@testing-library/react-hooks` wasn't being listed as a dev dependency, that's why ESLint was correct when complaining about it. With that being said, this rule should not be disabled.

```suggestion
```

File: .eslintrc.js [hunk 3]
```diff
@@ -9,15 +9,13 @@ module.exports = {
...
   },
   settings: {
     "import/resolver": {
       typescript: {},
       node: {
-        extensions: [
-          ".ts",
```
Comment (LauraBeatris) [inline .eslintrc.js]: I would suggest to not apply unrelated changes to the PR, and since it's just a style change that isn't being applied because of lint warnings or errors, can we rollback this?

File: .lintstagedrc.json [hunk 4]
```diff
@@ -1,6 +1,3 @@
 {
-  "*.{ts,tsx}": [
-    "yarn lint-code-style",
-    "git add"
-  ]
+  "*.{ts,tsx}": ["yarn lint-code-style", "yarn test", "git add"]
```
Comment (LauraBeatris) [inline .lintstagedrc.json]: I would suggest to not run the tests when adding changes to stage. It can slow down the development process since every time that a commit is made, it will take more than 30 seconds in order to execute all these scripts. 

```suggestion
  "*.{ts,tsx}": ["yarn lint-code-style", "git add"]
```

File: __tests__/useSelectedItems.test.tsx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+import useSelectedItems from "../src/index";
+
+const initialExampleItems = [
+  { id: 1, text: "What's up" },
+  { id: 2, text: "Hey there" },
+  { id: 3, text: "How you doin'" },
+];
+
+interface ExampleItem {
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: If `initialExampleItems` change, the interface will also need to be updated. But we can improve this in order to be extensible since the information is already available. 

Change the interface to be a literal type and access the type of any `initialExampleItems` element
```
type ExampleItem = typeof initialExampleItems[number]
```
image

File: __tests__/useSelectedItems.test.tsx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  id: number;
+  text: string;
+}
+
+describe("useSelectedItems", () => {
+  it("should return all items", () => {
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: Do you think it's better if we change this test description to `should initialize items` instead of `should return all items`?

By reading `should return all items`, I feel that `all items` may be generic.

File: __tests__/useSelectedItems.test.tsx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    const expectedItems = initialExampleItems.map((item) => ({
+      ...item,
+      isSelected: false,
+    }));
+
+    expect(result.current.items).toEqual(expectedItems);
+  });
+
+  it("should render with initialSelectedItems", () => {
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: Do you think it's better if we change this test description to `should initialize selected items` instead of `should render with initialSelectedItems`?

By reading `should render with initialSelectedItems`, I feel that it's related to rendering something but since the hook is a function, it returns the items, and does not render them.

File: __tests__/useSelectedItems.test.tsx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+];
+
+interface ExampleItem {
+  id: number;
+  text: string;
+}
+
+describe("useSelectedItems", () => {
+  it("should return all items", () => {
+    const { result } = renderHook(() => useSelectedItems<ExampleItem, ExampleItem["text"]>({
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I would suggest to extract `ExampleItem["text"]` to a type literal in order to reuse. It's going to be easier to change in the future if the identifier changes. Refer to:

```
type ExampleItemIdentifier = ExampleItem["text"]
```

And also, the identifier should be `"id"` instead of `"text"` since the `itemIdentifierKey` is being provided as `"id"`

File: __tests__/useSelectedItems.test.tsx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+      initialItems: initialExampleItems,
+    }));
+
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[0]),
+    );
+
+    expect(result.current.items).toContainEqual(
+      expect.objectContaining({ ...initialExampleItems[0], isSelected: true }),
+    );
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I would suggest to extract `initialExampleItems[0]` to a variable called `selectedItem`, since it can be reused for the assertions.

```suggestion
      initialSelectedItems: [selectedItem],
      initialItems: initialExampleItems,
    }));

    expect(result.current.selectedItems).toContainEqual(
      expect.objectContaining(selectedItem),
    );

    expect(result.current.items).toContainEqual(
      expect.objectContaining({ ...selectedItem, isSelected: true }),
    );
```

File: __tests__/useSelectedItems.test.tsx [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+  it("should toggle a single item", () => {
+    const { result } = renderHook(() => useSelectedItems<ExampleItem, ExampleItem["text"]>({
+      itemIdentifierKey: "id",
+      initialItems: initialExampleItems,
+    }));
+
+    act(() => {
+      result.current.toggleSingleItem(1);
+    });
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I would suggest to extract the elements that are going to be selected and reference in variables:

(I don't feel strong about the name of the variables bellow so feel free to use your judgement here)
```
const selectedItem1 = initialExampleItems[0];
const selectedItem2 = initialExampleItems[1];
```

And by doing so, these variables identifiers can be access bellow, without being to be hard coded:

```
 act(() => {
      result.current.toggleSingleItem(selectedItem1.id);
    });
 ```

File: __tests__/useSelectedItems.test.tsx [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    expect(result.current.selectedItems).toContainEqual(
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: This assertion is being repeated in the line 73. I would suggest to make all the assertion at the end of the testing block. 

```suggestion
```

File: __tests__/useSelectedItems.test.tsx [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    act(() => {
+      result.current.toggleSingleItem(2);
+    });
+
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[0]),
+    );
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[1]),
+    );
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I would suggest to "merge" these assertions since the two items can be verified at the same time.

```expect(result.current.selectedItems).toEqual([initialExampleItems[0], initialExampleItems[1]]);```

But this it's not going to work if the current implementation since I also noticed an error in the `selectedItems` implementation. It should't have the `isSelected` state being returned within the items since it's already selected... kind of redundant. 

In order to fix it, the selected items should be returned like the following:
```
const selectedItems = useMemo(() => (
    items
      .filter(item => item.isSelected)
      .map(({ isSelected: _isSelected, ...rest }) => ({
        ...rest,
      })) as T[]
  ), [items]);
  ```
  
I will make this change in another PR and then you can pull the changes here

File: __tests__/useSelectedItems.test.tsx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+  });
+
+  it("should toggle a single item", () => {
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I think that this test is being repeated above

File: __tests__/useSelectedItems.test.tsx [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+      expect.objectContaining(initialExampleItems[0]),
+    );
+
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[1]),
+    );
+
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[2]),
+    );
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I would suggest to "merge" the assertions by directly verifying if the initial items array is equal to the current `selectedItems` state

```suggestion
    expect(result.current.selectedItems).not.toEqual(initialExampleItems);

    act(() => {
      result.current.toggleAllItems();
    });

    expect(result.current.selectedItems).toEqual(initialExampleItems);
```

File: __tests__/useSelectedItems.test.tsx [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+    expect(() => {
+      renderHook(() => useSelectedItems<ExampleItem, ExampleItem["text"]>({
+        itemIdentifierKey: "wrong-id",
+        initialItems: initialExampleItems,
+      }));
+    }).toThrowError();
+
+    expect(consoleSpy).toHaveBeenCalled();
+    consoleSpy.mockRestore();
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: Is there an option in which we can restore all mocks after running each test?

File: __tests__/useSelectedItems.test.tsx [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+  it("shouldn't render if itemIdentifierKey is wrong", () => {
+    const consoleSpy = jest.spyOn(console, "error").mockImplementation();
+
+    expect(() => {
+      renderHook(() => useSelectedItems<ExampleItem, ExampleItem["text"]>({
+        itemIdentifierKey: "wrong-id",
+        initialItems: initialExampleItems,
+      }));
+    }).toThrowError();
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: This not seems to be working with the current version of `@testing-library/react-hooks` 

The hook result has an error property so it's easier to assert if an error happened or not.

```
  expect(result.error).toEqual(Error("Please, make sure to provide a valid identifier key to all items"));
```

I'll also suggest to create a constant for the error message and export it, so it can be reused in the hook module & other files.

File: __tests__/useSelectedItems.test.tsx [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    expect(result.current.selectedItems).toContainEqual(
+      expect.objectContaining(initialExampleItems[0]),
+    );
+
+    expect(result.current.items).toContainEqual(
+      expect.objectContaining({ ...initialExampleItems[0], isSelected: true }),
+    );
+  });
+
+  it("should render with empty items", () => {
```
Comment (LauraBeatris) [inline __tests__/useSelectedItems.test.tsx]: I don't know if it's a use case that we should support. The hook will be called referencing another array so it shouldn't be `undefined` when declaring it.

It can be `undefined` if it's a value that can be returned from a promise, with some kind of overhead, and that's why it's being coalesced to an empty array inside the hook module.

**Review by LauraBeatris — CHANGES_REQUESTED** [review]

File: __tests__/useSelectedItems.test.tsx (hunk 16 above)
Comment (LauraBeatris) [reply __tests__/useSelectedItems.test.tsx]: The error constant can be declared inside `constants.ts` and I would suggest something like that:

```
const ERRORS = {
 INVALID_UNIQUE_IDENTIFIER: "Please, make sure to provide a valid identifier key to all items"
}
```

But feel free to use your judgment here

File: src/index.ts [hunk 18]
```diff
@@ -45,7 +45,7 @@ function useSelectedItems<T extends DefaultItem, K extends string>({
   }, []);
 
   useEffect(() => {
-    const shouldInitializeItems = (initialItems ?? []).length > 0;
+    const shouldInitializeItems = initialItems.length > 0;
```
Comment (LauraBeatris) [inline src/index.ts]: https://github.com/LauraBeatris/use-selected-items-hook/pull/19#discussion_r571441820

File: src/types.ts [hunk 19]
```diff
@@ -3,7 +3,7 @@ export type DefaultItem = Record<any, any>;
 export type DefaultItemIdentifierKey = any;
 
 export interface Arguments<T = DefaultItem, K = DefaultItemIdentifierKey> {
-  initialItems: T[];
+  initialItems?: T[];
```
Comment (LauraBeatris) [inline src/types.ts]: https://github.com/LauraBeatris/use-selected-items-hook/pull/19#discussion_r571441820

File: __tests__/useSelectedItems.test.tsx (hunk 12 above)
Comment (LauraBeatris) [reply __tests__/useSelectedItems.test.tsx]: Changes to `selectedItems` applied at 4b52965826caf784ad47758f8c2e4d1edf549bb5

Now you can pull the changes from `master`

**Comment by LauraBeatris** [comment]
@pedrobslisboa 

Seems good! I've requested some changes to improve reusability, semantic and fix other issues. The library to test the hooks used in the first PR draft is deprecated, so we'll have to install the correct one and rollback the ESLint change to disable `import/no-extraneous-dependencies` since the library should be listed as a `devDependency`

I've refactored the way that `selectedItems` is being returned at 4b52965826caf784ad47758f8c2e4d1edf549bb5 and I'll release a new minor version after this PR be merged

File: __tests__/useSelectedItems.test.tsx (hunk 15 above)
Comment (LauraBeatris) [reply __tests__/useSelectedItems.test.tsx]: Jest has a configuration option called `resetMocks`. By enabling it, all mocks will be reset after running each test.

https://jestjs.io/docs/en/configuration#resetmocks-boolean

**Comment by LauraBeatris** [comment]
We should also test for the case where the items are asynchronously initialized

**Comment by LauraBeatris** [comment]
Hi @pedrobslisboa

My apologies but since this PR has been inactive, I'm going to fix the improvements that I mentioned bellow and merge it. Thanks for the contribution 🙂

**Comment by LauraBeatris** [comment]
Besides testing the component interface, we should also add a unit test for the reducer logic
